### DISCUSSION TITLE
How does jax implement jax.lax.psum and jax.lax.all_to_all?

### DISCUSSION THREAD

**Discussion by anirudhitagi** [jax-ml/jax#18018] [Q&A]
Hello all,

I am new to Jax and I am exploring collective communications and so I wish to understand how does Jax implement `jax.lax.psum`, `jax.lax.all_gather`  and `jax.lax.all_to_all` ?

Is it implemented as Ring, Binary tree, halving-doubling or something else? 
And does this implementation depend on the hardware it is being deployed on? For context, I am experimenting with TPU v-3s. 

Any documentation or explanation would be most appreciated !

Thanks and Regards,
AI

**Comment by ayaka14732** [comment]
You can now implement them using your desired algorithm in Pallas. For example, for a TPU v4-16 (2×2×2) all-reduce program transferring 96 KiB per chip, my algorithm implemented in Pallas TPU achieved a 50.47% performance improvement over a standard XLA program.
